python/ray/data/_internal/planner/plan_read_op.py

```diff
@@ -93,25 +93,24 @@ def plan_read_op(op: Read) -> PhysicalOperator:
     See Planner.plan() for more details.
     """
 
-    def get_input_data() -> List[RefBundle]:
-        read_tasks = op._reader.get_read_tasks(op._parallelism)
-        return [
-            RefBundle(
-                [
-                    (
-                        # TODO(chengsu): figure out a better way to pass read
-                        # tasks other than ray.put().
-                        ray.put(read_task),
-                        cleaned_metadata(read_task),
-                    )
-                ],
-                owns_blocks=True,
-            )
-            for read_task in read_tasks
-        ]
+    read_tasks = op._reader.get_read_tasks(op._parallelism)
+    input_data = [
+        RefBundle(
+            [
+                (
+                    # TODO(chengsu): figure out a better way to pass read
+                    # tasks other than ray.put().
+                    ray.put(read_task),
+                    cleaned_metadata(read_task),
+                )
+            ],
+            owns_blocks=True,
```

Comment:
If you set owns_blocks=False that would also equivalently prevent them from being freed.

Reply:
Though, it sounds like we'd like to free them at the end of the workload, just not as soon as possible, so there is some subtlety here.

Reply:
yes, `owns_blocks=False` is the right fix. at the end of the workload, they will be eventually freed by python GC. and their sizes are small. So it should be fine.